File: mobile/android/app/src/main/java/com/remotedesktop/config/FirebaseConfig.kt
package com.remotedesktop.config

import org.json.JSONObject

data class FirebaseConfig(
    val apiKey: String,
    val applicationId: String,
    val projectId: String,
    val databaseURL: String
) {
    companion object {
        const val PAYLOAD_VERSION = 1

        fun fromQrPayload(raw: String): FirebaseConfig? {
            return try {
                val obj = JSONObject(raw)
                if (obj.optInt("v", -1) != PAYLOAD_VERSION) return null
                if (obj.optString("backend") != "firebase") return null
                val android = obj.optJSONObject("android") ?: return null
                FirebaseConfig(
                    apiKey = android.optString("apiKey").takeIf { it.isNotBlank() } ?: return null,
                    applicationId = android.optString("applicationId").takeIf { it.isNotBlank() } ?: return null,
                    projectId = android.optString("projectId").takeIf { it.isNotBlank() } ?: return null,
                    databaseURL = android.optString("databaseURL").takeIf { it.isNotBlank() } ?: return null
                )
            } catch (_: Exception) {
                null
            }
        }
    }

    fun toJson(): String =
        JSONObject().apply {
            put("apiKey", apiKey)
            put("applicationId", applicationId)
            put("projectId", projectId)
            put("databaseURL", databaseURL)
        }.toString()
}

File: mobile/ios/RemoteDesktop/Sources/FirebaseConfig.swift
import Foundation

struct FirebaseConfig: Equatable, Codable {
    let apiKey: String
    let googleAppId: String
    let projectId: String
    let databaseURL: String
    let bundleId: String

    static let payloadVersion = 1

    static func fromQrPayload(_ raw: String) -> FirebaseConfig? {
        guard let data = raw.data(using: .utf8) else { return nil }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        guard let v = json["v"] as? Int, v == payloadVersion else { return nil }
        guard let backend = json["backend"] as? String, backend == "firebase" else { return nil }
        guard let ios = json["ios"] as? [String: Any] else { return nil }
        guard
            let apiKey = (ios["apiKey"] as? String)?.trimmedNonEmpty(),
            let googleAppId = (ios["googleAppId"] as? String)?.trimmedNonEmpty(),
            let projectId = (ios["projectId"] as? String)?.trimmedNonEmpty(),
            let databaseURL = (ios["databaseURL"] as? String)?.trimmedNonEmpty(),
            let bundleId = (ios["bundleId"] as? String)?.trimmedNonEmpty()
        else { return nil }
        return FirebaseConfig(
            apiKey: apiKey,
            googleAppId: googleAppId,
            projectId: projectId,
            databaseURL: databaseURL,
            bundleId: bundleId
        )
    }
}

private extension String {
    func trimmedNonEmpty() -> String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
